### PR TITLE
API Guidelines blueprint

### PR DESCRIPTION
**Pull request by ocket8888**
## What does this PR (Pull Request) do?

- [x] This PR is not related to any Issue

Adds a blueprint for some API guidelines.

## Which Traffic Control components are affected by this PR?
- Documentation

## What is the best way to verify this PR?
Read the blueprint. Maybe pass it through a Markdown parser/validator/renderer if you want.

## The following criteria are ALL met by this PR
- [x] Tests are unnecessary
- [x] This PR includes documentation
- [x] An update to CHANGELOG.md is not necessary
- [x] This PR includes any and all required license headers
- [x] This PR does not include a database migration
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY**

### CODE REVIEW

File: blueprints/api.guidelines.md [hunk 1]
```diff
@@ -0,0 +1,535 @@
...
+from being fulfilled via one or more `error`-level alerts.
+
+#### 500 Internal Server Error
+When the Traffic Ops server encounters some error - through no fault of the
+client or their request - that renders it incapable of servicing the client's
+request, it MUST return this status code if no other code is more appropriate.
+The response body in this case SHOULD indicate that an error occurred, but
+MUST NOT divulge details about what data was being processed, what (if any)
+other components are not functioning properly, or what process failed. Generally
+it is advisable that the resultant \code{alerts} array entry just say "Internal
```
Comment (mitchell852) [inline blueprints/api.guidelines.md]: do you need to state that the details of the error (if known) should be written to TO logs instead?

File: blueprints/api.guidelines.md [hunk 2]
```diff
@@ -0,0 +1,535 @@
...
+Server Error" and nothing else.
+
+#### 501 Not Implemented}
```
Comment (mitchell852) [inline blueprints/api.guidelines.md]: curly

File: blueprints/api.guidelines.md [hunk 3]
```diff
@@ -0,0 +1,535 @@
...
+to be passed in the query string to separate it from the object definition.
+- The following query parameters are reserved for special use by Traffic Ops
+endpoint handlers, and may not be used for any purpose other than their
+prescribed functions.
+
+	- `limit`
+	- `newerThan`
+	- `offset`
+	- `olderThan`
+	- `orderby`
```
Comment (mitchell852) [inline blueprints/api.guidelines.md]: should this be `orderBy`?

File: blueprints/api.guidelines.md [hunk 4]
```diff
@@ -0,0 +1,535 @@
...
+DELETE destroys an object stored on the server. Typically the request will
+contain identifying information for the object(s) to be destroyed either in the
+request URI or in the request's body. Traffic Ops API endpoint authors MUST use
+this request method whenever an object identified by the request URI is being
+destroyed. When such deletion successfully occurs, the response body MUST
+contain a representation of the destroyed object.
+
+### HTTP Response Codes
+Proper use of HTTP response codes can significantly improve user interfaces
+built on top of the API. What follows is a (non-exhaustive) set of response
```
[the remark this answers is not in the document]
Comment (ocket8888) [reply blueprints/api.guidelines.md]: Nah, I don't want to limit the API's growth. A new response code could be standardized at any time, so even if these are sufficient out of the ones that exist today, they aren't representative of all possible response codes for all time.

File: blueprints/api.guidelines.md [hunk 5]
```diff
@@ -0,0 +1,535 @@
...
+access to the resource they are requesting. It MUST also be used when they have
+some access to the resource, but not with the specific request method they used.
+This can pertain to restricted access on the basis of Role, User Permissions, as
+well as Tenancy.
+
+The response body MUST NOT disclose any information regarding why the user was
+denied access.
+
+#### 404 Not Found
+This MUST be the returned status code when the client requests a path that does
```
[the remark this answers is not in the document]
Comment (ocket8888) [reply blueprints/api.guidelines.md]: I suppose I can, though I think that's part of the definition of a "path" part of a URL.

File: blueprints/api.guidelines.md (hunk 3 above)
Comment (ocket8888) [reply blueprints/api.guidelines.md]: nope. Our current API uses "orderby," unfortunately. Changing that would be a breaking change unless they both worked - and I don't think anyone wants to bother with that. I know I don't. So we'll save consistent casing for v3 (or even v4 maybe, at this point).

File: blueprints/api.guidelines.md (hunk 1 above)
Comment (ocket8888) [reply blueprints/api.guidelines.md]: I would hope not. This is just about the API itself.

File: blueprints/api.guidelines.md [hunk 6]
```diff
@@ -0,0 +1,549 @@
...
+Endpoints MUST use the following, reserved properties of `summary` for their
+described purposes (when use of `summary` is appropriate) rather than defining
+new `summary` or `response` properties to suit the same purpose:
+
+- `count` - Count contains an unsigned integer that defines the total number of
+results that could possibly be returned given the non-pagination query
+parameters supplied by the client.
+
+### HTTP Request Methods
+[RFC 2616 Section 9](https://tools.ietf.org/html/rfc2616#section-9) defines the
```
Comment (rob05c) [inline blueprints/api.guidelines.md:141]: RFC 2616 is obsolete, should be RFC 7231 Section 4: https://tools.ietf.org/html/rfc7231#section-4

**Pull request merged**